whats-that-typeclass-Foldable/article.md

```diff
@@ -0,0 +1,422 @@
+
+```haskell
+foldMap :: (Monoid m, Foldable t) => (a -> m) -> t a -> m
+```
+
+It's important to note that `foldMap` has a [`Monoid`](https://hackage.haskell.org/package/base-4.16.0.0/docs/Prelude.html#t:Monoid) constraint. Its first argument is a function that maps each element of a container into a monoid, the second argument is a container itself. After mapping elements, the results are combined using `(<>)` operator. The order of folding is from right to left, so `foldMap` could be implemented via `foldr`.
+
+Let's look at how exactly that could be done:
+
+```haskell
+foldMap :: (Monoid m, Foldable t) => (a -> m) -> t a -> m
+foldMap f = foldr (\x acc -> f x <> acc) mempty
+```
+
+`foldMap` doesn't have a base element since only the elements of the container are reduced. However, `foldr` does, so it perfectly makes sense to use the identity of monoid — `mempty`. We can also see that the folding function `f` is composed with `(<>)`, thus, the current result is appended to a monoid accumulator on each step.
+
+Recall how the addition of list elements is implemented with `foldr`:
+
+```haskell
+ghci> foldr (+) 0 [1, 2, 3]
+6
+```
+
+To do the same in terms of monoid, consider a monoid under addition — [`Sum`](https://hackage.haskell.org/package/base-4.16.0.0/docs/Data-Monoid.html#t:Sum). We could use `foldMap` and `Sum` to perform the addition of list elements:
+
+```haskell
+-- import `Data.Monoid` to get `Sum`
+ghci> import Data.Monoid
+
+ghci> foldMap Sum [1, 2, 3]
+Sum {getSum = 6}
+```
+
+Note that here constructor `Sum` is a function that turns a list element into monoid. As expected, we get the same result wrapped in `Sum` that can easily be accessed via `getSum`.
+
+To conclude, `foldMap` and `foldr` are interchangeable, the difference is that the former receives the combiner and the base element from the `Monoid` instance, and the latter accepts them explicitly.
+
+### Minimal complete definition
+
+We have already shown how `foldMap` is implemented using `foldr`. It's not obvious, but `foldr` could also be implemented via `foldMap`! It's off-topic for this article, therefore, please proceed to [the documentation](https://hackage.haskell.org/package/base-4.16.0.0/docs/src/Data.Foldable.html#foldr) for the details. Consequently, to create the `Foldable` instance, you can provide a definition of either `foldr` or `foldMap`, which is exactly the minimal complete definition — `foldMap | foldr`. Note that you shouldn't implement `foldMap` in terms of `foldr` and simultaneously `foldr` in terms of `foldMap` as this will just loop forever. Thus, you implement one of them, and Haskell provides the definitions of all `Foldable`'s methods automatically.
+
+## Other methods of `Foldable` and things to be careful about when using the type class
+
+`foldl`, `foldr`, and `foldMap` could be considered the core for understanding the nature of `Foldable`. Nonetheless, there are others which are worth mentioning.
+
+### Strict `foldl'`
```

Comment:
A bit of a small thing, but foldl' is strict, as I understand, so saying strict foldl` doesn't really make a lot of sense, right?